### ISSUE TITLE
No detections when i used the custom yaml file 

### ISSUE THREAD

**Issue by anazkhan** [ultralytics/ultralytics#13769]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am using custom yaml file inorder to change the activation functions to relu . but i am seeing no detections when i am using the same yaml file for inference . also i would like to know how can i change the activation functions for specififc layers. it will be great if you can explain in detail.

below is my yaml file.

# Ultralytics YOLO 🚀, AGPL-3.0 license
# YOLOv8 object detection model with P3-P5 outputs. For Usage examples see https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 80 # number of classes
activation: nn.ReLU()
scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024] # YOLOv8n summary: 225 layers,  3157200 parameters,  3157184 gradients,   8.9 GFLOPs
  s: [0.33, 0.50, 1024] # YOLOv8s summary: 225 layers, 11166560 parameters, 11166544 gradients,  28.8 GFLOPs
  m: [0.67, 0.75, 768] # YOLOv8m summary: 295 layers, 25902640 parameters, 25902624 gradients,  79.3 GFLOPs
  l: [1.00, 1.00, 512] # YOLOv8l summary: 365 layers, 43691520 parameters, 43691504 gradients, 165.7 GFLOPs
  x: [1.00, 1.25, 512] # YOLOv8x summary: 365 layers, 68229648 parameters, 68229632 gradients, 258.5 GFLOPs

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]] # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]] # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]] # 7-P5/32
  - [-1, 3, C2f, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]] # 9

# YOLOv8.0n head
head:
  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 6], 1, Concat, [1]] # cat backbone P4
  - [-1, 3, C2f, [512]] # 12

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 3, C2f, [256]] # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]] # cat head P4
  - [-1, 3, C2f, [512]] # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]] # cat head P5
  - [-1, 3, C2f, [1024]] # 21 (P5/32-large)

  - [[15, 18, 21], 1, Detect, [nc]] # Detect(P3, P4, P5)
  

model = YOLO("custom_config.yaml").load("yolov8s.pt")

this is how i load the model.

**Comment by glenn-jocher** [comment]
Hi there,

Thank you for reaching out and providing your custom YAML file. Let's address your issue step-by-step.

### 1. Reproducible Example
To help us investigate the issue effectively, could you please provide a minimum reproducible code example? This will allow us to replicate the problem on our end. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

### 2. Verify Package Versions
Ensure that you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using the following commands:
```bash
pip install --upgrade torch ultralytics
```

### 3. Custom YAML File Adjustments
It appears that you are trying to modify the activation functions in the model. However, the `activation` parameter in the YAML file is not directly used to change the activation functions in specific layers. Instead, you need to modify the model architecture directly in the code.

Here's an example of how you can change the activation function for specific layers in the model:

```python
from ultralytics import YOLO
import torch.nn as nn

# Load the model
model = YOLO("yolov8s.pt")

# Modify the activation function for specific layers
for layer in model.model.modules():
    if isinstance(layer, nn.Conv2d):
        layer.activation = nn.ReLU()

# Save the modified model
model.save("custom_yolov8s.pt")
```

### 4. No Detections Issue
If you are experiencing no detections, it could be due to several reasons:
- The model might not be properly trained with the new configuration.
- There might be an issue with the custom activation functions affecting the model's performance.

To debug this, you can start by checking if the model is able to detect objects with the default configuration before making any modifications. If it works, then incrementally apply your changes to identify the exact cause.

### 5. Detailed Explanation
For a detailed explanation on how to modify specific layers and their activation functions, you can refer to the [YOLOv8 documentation](https://docs.ultralytics.com/models/yolov8/). This will provide you with comprehensive insights into the model architecture and customization options.

Feel free to share any additional details or code snippets that might help us understand the issue better. We're here to assist you!

**Comment by anazkhan** [comment]
Thank you for the reply,
I  would like to let you know that i have used exactly the same yaml file in official yolov8 the only modification made was setting activation argument to relu but irrespective of that even when i used the same yaml file but as custom loaded it was showing no detection during inference. I would like to know if there is any issue with the way i load the model.

model = YOLO("custom_config.yaml").load("yolov8s.pt")

**Comment by glenn-jocher** [comment]
Hi there,

Thank you for the additional details! Let's address your concerns step-by-step.

### Minimum Reproducible Example
To help us investigate effectively, could you please provide a minimum reproducible code example? This will allow us to replicate the issue on our end. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

### Verify Package Versions
Ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using:
```bash
pip install --upgrade torch ultralytics
```

### Custom YAML File and Model Loading
The issue might be related to how the model is being loaded with the custom YAML file. Here’s a more detailed approach to ensure the activation functions are properly set:

```python
from ultralytics import YOLO
import torch.nn as nn

# Load the model with the custom YAML configuration
model = YOLO("custom_config.yaml").load("yolov8s.pt")

# Modify the activation function for specific layers
for layer in model.model.modules():
    if isinstance(layer, nn.Conv2d):
        layer.activation = nn.ReLU()

# Save the modified model
model.save("custom_yolov8s.pt")

# Run inference to check detections
results = model("path/to/your/image.jpg")
results.show()
```

### Debugging No Detections
If the model still shows no detections, try the following steps:
1. **Baseline Check**: Ensure the model works with the default configuration before any modifications.
2. **Incremental Changes**: Apply changes incrementally to identify the exact cause.
3. **Model Training**: Verify if the model needs retraining with the new configuration.

Feel free to share any additional details or code snippets that might help us understand the issue better. We're here to assist you! 😊
